### DISCUSSION TITLE
Counting newlines in output

### DISCUSSION THREAD

**Discussion by tsoutsman** [BurntSushi/ripgrep#2176] [General] [answered]
Hi, this question doesn't strictly relate to ripgrep, but hopefully, it's not too off-topic.

The command `rg example | wc -l` counts the number of matches. According to wc documentation, `-l` counts the number of newlines, but the ripgrep output clearly contains more lines than matches (whitespace, file names, etc.). How does `wc -l` still correctly return the number of matches?

**Comment by BurntSushi** [comment, marked as answer]
Try comparing the output of `rg example` and `rg example | cat`. That should hopefully answer your question.

Try the same with `ls` and `ls | cat` for example.

The trick here is that tools like ripgrep and ls change their output format based on whether they're writing to a tty or not.
